### PR TITLE
Display qr code in tag properties

### PR DESCRIPTION
**Pull request by thomasloven**
## Proposed change

Display tags as QR codes in properties
image

It's an `` so it can be downloaded or saved to a phone.

Unfortunately the contrast seems to be too low to reliably scan if it's blue instead of black.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/config/tags/dialog-tag-detail.ts [hunk 1]
```diff
@@ -191,13 +195,43 @@ class DialogTagDetail extends LitElement implements HassDialog {
     }
   }
 
+  private async _generateQR() {
+    let canvas = await QRCode.toCanvas(
```
Comment (zsarnett) [inline src/panels/config/tags/dialog-tag-detail.ts:199]:
```suggestion
    const canvas = await QRCode.toCanvas(
```

File: src/panels/config/tags/dialog-tag-detail.ts [hunk 2]
```diff
@@ -191,13 +195,43 @@ class DialogTagDetail extends LitElement implements HassDialog {
...
+      `https://home-assistant.io/tag/${this._params?.entry?.id}`,
+      {
+        width: 180,
+        errorCorrectionLevel: "Q",
+      }
+    );
+    let context = canvas.getContext("2d");
```
Comment (zsarnett) [inline src/panels/config/tags/dialog-tag-detail.ts:206]:
```suggestion
    const context = canvas.getContext("2d");
```

File: src/panels/config/tags/dialog-tag-detail.ts [hunk 3]
```diff
@@ -191,13 +195,43 @@ class DialogTagDetail extends LitElement implements HassDialog {
...
+
+    let imageObj = new Image();
```
Comment (zsarnett) [inline src/panels/config/tags/dialog-tag-detail.ts:208]:
```suggestion
    const imageObj = new Image();
```

File: src/panels/config/tags/dialog-tag-detail.ts [hunk 4]
```diff
@@ -191,13 +195,43 @@ class DialogTagDetail extends LitElement implements HassDialog {
...
+    imageObj.src = "/static/icons/favicon-192x192.png";
```
Comment (zsarnett) [inline src/panels/config/tags/dialog-tag-detail.ts:209]: We should probably make this a const somewhere. Atleast top of the file? could definitely change

File: src/panels/config/tags/dialog-tag-detail.ts [hunk 5]
```diff
@@ -191,13 +195,43 @@ class DialogTagDetail extends LitElement implements HassDialog {
...
+    await new Promise((resolve) => (imageObj.onload = resolve));
```
Comment (zsarnett) [inline src/panels/config/tags/dialog-tag-detail.ts:210]:
```suggestion
    await new Promise((resolve) => {(imageObj.onload = resolve)});
```

File: src/panels/config/tags/dialog-tag-detail.ts (hunk 1 above)
Comment (thomasloven) [reply src/panels/config/tags/dialog-tag-detail.ts]: I cannot understand why eslint doesn't tell me about this when I run it locally for the life of me!

File: src/panels/config/tags/dialog-tag-detail.ts (hunk 4 above)
Comment (thomasloven) [reply src/panels/config/tags/dialog-tag-detail.ts]: There could probably be a `src/resources/static.ts`or something to keep all of this kind of stuff, but there really aren't that many as it turns out:
```
~/c/h/frontend > git grep "/static/" src/
src/common/dom/setup-leaflet-map.ts:  Leaflet.Icon.Default.imagePath = "/static/images/leaflet/images/";
src/common/dom/setup-leaflet-map.ts:  style.setAttribute("href", "/static/images/leaflet/leaflet.css");
src/components/ha-icon.ts:    const iconPromise = fetch(`/static/mdi/${chunk}.json`).then((response) =>

src/html/_header.html.template:<link rel='icon' href='/static/icons/favicon.ico'>
# ... html templates (removed a couple of lines)

src/layouts/ha-init-page.ts:        <img src="/static/icons/favicon-192x192.png" height="192" />
src/panels/config/info/ha-config-info.ts:              src="/static/icons/favicon-192x192.png"
src/panels/config/tags/dialog-tag-detail.ts:const QR_LOGO_URL = "/static/icons/favicon-192x192.png";
src/panels/lovelace/cards/hui-media-control-card.ts:        background-image: url("../static/images/card_media_player_bg.png");
src/panels/lovelace/components/hui-image.ts:        background: grey url("/static/images/image-broken.svg") center/36px

src/resources/roboto.js:  url(/static/fonts/roboto/Roboto-Thin.woff2) format("woff2");
# ... font stuff (removed a couple of lines)...

src/util/hass-translation.ts:  const response = await fetch(`/static/translations/${fingerprint}`, {
```

The media-control-card css is an odd one out, though. Expect a PR.

**Comment by bramkragten** [comment]
What is the idea behind this? That a user downloads the QR code, prints it, and scans it with their phone to trigger a command?

I think we should add some instructions and clarification, now someone may think it is just QR or something.

**Comment by thomasloven** [comment]
Good point

**Comment by thomasloven** [comment]
Any suggestions on wording?
image

**Comment by bramkragten** [comment]
Align the text to the right.

Not sure how heavy the QR code lib is, but can we create the QR code on the click of a button instead of always when opening the dialog? And also load the resources for creating the QR code then? We could show a dummy QR code you could click to create one?

Text maybe something like:

> A tag can trigger an automation when scanned, you can use NFC tags, QR codes, or any other kind of tag.
> Use our [companion apps](#) to write this tag to a programmable NFC tag or create a QR code below.

**Comment by thomasloven** [comment]
Is there any good way to put links in translated strings?

**Comment by bramkragten** [comment]
Yes, you can just pass a `html``` as a var in lokalise that contains another translation.

**Comment by thomasloven** [comment]
image
image

File: src/panels/config/tags/dialog-tag-detail.ts [hunk 6]
```diff
@@ -191,13 +234,43 @@ class DialogTagDetail extends LitElement implements HassDialog {
...
+    });
+    context.drawImage(
+      imageObj,
+      canvas.width / 3,
+      canvas.height / 3,
+      canvas.width / 3,
+      canvas.height / 3
+    );
+
+    this._qrCode = html`<img src=${canvas.toDataURL()}></img>`;
```
Comment (bramkragten) [inline src/panels/config/tags/dialog-tag-detail.ts:261]: We should set `_qrCode` to undefined when the dialog is closed.

**Review by bramkragten — APPROVED** [review]

File: src/panels/config/tags/dialog-tag-detail.ts (hunk 6 above)
Comment (thomasloven) [reply src/panels/config/tags/dialog-tag-detail.ts:261]: It's set to undefined when the dialog is opened. Both? Or only on close?

Comment (bramkragten) [reply src/panels/config/tags/dialog-tag-detail.ts:261]: On close then, as it saves memory.

**Review by bramkragten — APPROVED** [review]

**Pull request merged**